tests/t_list_test.cc

```diff
@@ -101,6 +114,117 @@ TEST_F(RedisListTest, Range) {
   list->Del(key_);
 }
 
+TEST_F(RedisListTest, Rem) {
+  int ret;
+  uint32_t len;
+  list->Push(key_, fields_, false, &ret);
+  EXPECT_EQ(fields_.size(), ret);
+  Slice del_elem("list-test-key-1");
+  // lrem key_ 1 list-test-key-1
+  list->Rem(key_, 1, del_elem, &ret);
+  EXPECT_EQ(1, ret);
+  list->Size(key_, &len);
+  EXPECT_EQ(fields_.size()-1, len);
+  for (size_t i = 1; i < fields_.size(); i++) {
+    std::string elem;
+    list->Pop(key_, &elem, true);
+    EXPECT_EQ(elem, fields_[i].ToString());
+  }
+  // lrem key_ 0 list-test-key-1
+  list->Push(key_, fields_, false, &ret);
+  EXPECT_EQ(fields_.size(), ret);
+  list->Rem(key_, 0, del_elem, &ret);
+  EXPECT_EQ(4, ret);
+  list->Size(key_, &len);
+  EXPECT_EQ(fields_.size()-4, len);
+  for (size_t i = 0; i < fields_.size(); i++) {
+    std::string elem;
+    if (fields_[i] == del_elem) continue;
+    list->Pop(key_, &elem, true);
+    EXPECT_EQ(elem, fields_[i].ToString());
+  }
+  // lrem key_ 1 nosuchelement
+  Slice no_elem("no_such_element");
+  list->Push(key_, fields_, false, &ret);
+  EXPECT_EQ(fields_.size(), ret);
+  list->Rem(key_, 1, no_elem, &ret);
+  EXPECT_EQ(0, ret);
+  list->Size(key_, &len);
+  EXPECT_EQ(fields_.size(), len);
+  for (size_t i = 0; i < fields_.size(); i++) {
+    std::string elem;
+    list->Pop(key_, &elem, true);
+    EXPECT_EQ(elem, fields_[i].ToString());
+  }
+  // lrem key_ -1 list-test-key-1
+  list->Push(key_, fields_, false, &ret);
+  list->Rem(key_, -1, del_elem, &ret);
+  EXPECT_EQ(1, ret);
+  list->Size(key_, &len);
+  EXPECT_EQ(fields_.size()-1, len);
+  int cnt = 0;
+  for (size_t i = 0; i < fields_.size(); i++) {
+    std::string elem;
+    if (fields_[i] == del_elem) {
+      if (++cnt > 3) continue;
+    }
+    list->Pop(key_, &elem, true);
+    EXPECT_EQ(elem, fields_[i].ToString());
+  }
+  // lrem key_ -5 list-test-key-1
+  list->Push(key_, fields_, false, &ret);
+  EXPECT_EQ(fields_.size(), ret);
+  list->Rem(key_, -5, del_elem, &ret);
+  EXPECT_EQ(4, ret);
+  list->Size(key_, &len);
+  EXPECT_EQ(fields_.size()-4, len);
+  for (size_t i = 0; i < fields_.size(); i++) {
+    std::string elem;
+    if (fields_[i] == del_elem) continue;
+    list->Pop(key_, &elem, true);
+    EXPECT_EQ(elem, fields_[i].ToString());
+  }
+  list->Del(key_);
+}
+
+TEST_F(RedisListSpecificTest, Rem) {
+  int ret;
+  list->Push(key_, fields_, false, &ret);
+  EXPECT_EQ(fields_.size(), ret);
+  Slice del_elem("9");
+  // lrem key_ 1 9
+  list->Rem(key_, 1, del_elem, &ret);
+  EXPECT_EQ(1, ret);
+  uint32_t len;
+  list->Size(key_, &len);
+  EXPECT_EQ(fields_.size()-1, len);
+  int cnt = 0;
+  for (size_t i = 0; i < fields_.size(); i++) {
+    if (fields_[i] == del_elem) {
+      if (++cnt <= 1) continue;
+    }
+    std::string elem;
+    list->Pop(key_, &elem, true);
+    EXPECT_EQ(elem, fields_[i].ToString());
```

Comment:
Actually I prefer we can also add some tcl test cases, you can show the new list members for comparison, that may be easier and clearer

Reply:
Yes, tcl test cases can be easier and clearer. I will add tcl cases for `lrem` and `ltrim` later.